Mathlib/Tactic/Find.lean

```diff
@@ -39,96 +102,249 @@ private partial def matchHyps : List Expr → List Expr → List Expr → MetaM
+
+
+/-- Parses a list of `find_pattern` syntax into `Arguments` -/
+def parseFindPatterns (args : TSyntaxArray ``find_pattern) : TermElabM Arguments := do
+  let mut idents := #[]
+  let mut name_pats := #[]
+  let mut terms := #[]
+  for arg in args do
+    match arg with
+    | `(find_pattern| $ss:str) => do
+      let str := Lean.TSyntax.getString ss
+      name_pats := name_pats.push str
+    | `(find_pattern| $i:ident) => do
+      let n := Lean.TSyntax.getId i
+      unless (← getEnv).contains n do
+        throwErrorAt i "unknown identifier '{n}'"
+      idents := idents.push n
+    | `(find_pattern| $_:turnstyle $s:term) => do
+      let t ← Lean.Elab.Term.elabTerm s none
+      terms := terms.push (true, t)
+    | `(find_pattern| $s:term) => do
+      let t ← Lean.Elab.Term.elabTerm s none
+      terms := terms.push (false, t)
+    | _ => throwErrorAt arg "unexpected argument to #find"
+  pure {idents, name_pats, terms}
+
+open Command
+
+/--
+The `#find` command finds definitions and lemmas by various ways, which can be combined:
+
+1. By constant:
+   ```lean
+   #find Real.sin tsum
+   ```
+   finds all lemmas that somehow mention the sine function and an infinite series sum.
+
+2. By lemma name substring:
+   ```lean
+   #find tsum "two"
+   ```
+   finds all lemmas about `tsum` that have `"two"` as part of the lemma _name_.
+
+3. By subexpression:
+   ```lean
+   #find tsum (_ * _ ^ _)
+   ```
+   finds all lemmas that mention `tsum` and where somewhere in the lemma statement there is a
+   product with a power as the second argument. The pattern can also be non-linear, as in
+   ```lean
+   #find (Real.sqrt ?a * Real.sqrt ?a)
+   ```
+
+4. By conclusion:
+   ```lean
+   #find ⊢ (tsum _ = _ * tsum _)
+   ```
+   finds all lemmas where the conclusion (the subexpression to the right of all `→` and `∀`) has the
+   given shape. If the pattern has hypotheses, then they are matched against the hypotheses of
+   the lemma in any order:
+   ```lean
+   #find ⊢ (_ < _ → tsum _ < tsum _)
+   ```
+   will find `tsum_lt_tsum` even though the hypotheses `f i < g i` is not the last.
+
+The command will show the list of results in the info view, where you can cover to see theirs types
+or click to go to their definition.
+
+If you pass more than one such search filter, it will list lemmas matching _all_ of them. At least
+some filter must mention a concrete name, because `#find` maintains an index of which lemmas
+mention which other constants. This is also why the _first_ use of `#find` will be somewhat slow
+(typically less than half a minute with all of `Mathlib` imported), but subsequent uses are faster.
+
+It may be useful to open a scratch file, `import Mathlib`, and use `#find` there, this way you will
+find lemmas that you have not yet imported, and the cache will stay up-to-date.
+
+Inside tactic proofs, the `#find` tactic can be used instead.
+-/
```

Comment:
```suggestion
/--
The `#find` command finds definitions and lemmas in various ways. One can search by: the constants 
involved in the type; a substring of the name; a subexpression of the type; or a subexpression 
located in the return type or a hypothesis specifically. All of these search methods can be 
combined in a single query.

1. By constant:
   ```lean
   #find Real.sin
   ```
   finds all lemmas whose statement somehow mentions the sine function.

2. By lemma name substring:
   ```lean
   #find "two"
   ```
   finds all lemmas that have `"two"` as part of the lemma _name_.

3. By subexpression:
   ```lean
   #find (_ * (_ ^ _))
   ```
   finds all lemmas whose statements somewhere include a product where the second argument is 
   raised to some power. The pattern can also be non-linear, as in
   ```lean
   #find (Real.sqrt ?a * Real.sqrt ?a)
   ```

4. By conclusion and/or hypothesis:
   ```lean
   #find ⊢ (tsum _ = _ * tsum _)
   ```
   finds all lemmas where the conclusion (the subexpression to the right of all `→` and `∀`) has the
   given shape. If the pattern has hypotheses, they are matched against the hypotheses of
   the lemma in any order; for example,
   ```lean
   #find ⊢ (_ < _ → tsum _ < tsum _)
   ```
   will find `tsum_lt_tsum` even though the hypothesis `f i < g i` is not the last.
5. In combination:
   ```lean
   #find Real.sin "two" tsum  (_ * _) (_ ^ _) ⊢ (_ < _ → _)
   ```
   will find all lemmas which mention the constants `Real.sin` and `tsum`, have `"two"` as a 
   substring of the lemma name, include a product and a power somewhere in the type, *and* have a 
   hypothesis of the form `_ < _`. 

If you pass more than one such search filter, `#find` will only return lemmas which match _all_ of 
them simultaneously. At least some filter must mention a concrete name, because `#find` maintains 
an index of which lemmas mention which other constants. This is also why the _first_ use of `#find` 
will be somewhat slow (typically less than half a minute with all of `Mathlib` imported), but 
subsequent uses are faster.

It may be useful to open a scratch file, `import Mathlib`, and use `#find` there, this way you will
find lemmas that you have not yet imported, and the cache will stay up-to-date.

Inside tactic proofs, the `#find` tactic can be used instead.
-/
```
As someone who's never used `#find`, I found this docstring a bit confusing—was `#find` always supposed to take two arguments? My suggestion here is to separate the demonstration of each individual functionality from the demonstration that they can all be combined. I also fixed some typos and summarized the functionalities at the start of the docstring.

However, I'm not sure what a good candidate would be for the combination demonstration! There's probably something better than `#find Real.sin "two" tsum  (_ * _) (_ ^ _) ⊢ (_ < _ → _)`—I'm not sure if this even finds any lemmas. Someone more familiar with the specifics of Mathlib could probably come up with a good example where each filter is reasonable.

Whatever the example, I think it should (1) involve all search methods (2) involve at least one of the search methods *twice*, to show that you can do that, and preferably not in the same order—e.g. here we search for constants `Real.sin` and `tsum`, and do so with `"two"` in between.

Also, this is assuming I've understood the behavior correctly! (E.g., we only search the type of lemmas—not proof terms—for constants, right?)

Reply:
> I think we should try to get this working before merging as long as it's not too difficult—if it does prove difficult, it can wait for another PR. :)

Easy to do (I have it on a branch), harder if you want it to be snappy.

Reply:
I applied your suggestions, but reopened this conversation so that nothing is forgotten.